RFC: Cooperative Heaters predictive control

nefelim4ag opened the pull request:
Some time ago [I dug heaters a little on the Discourse](https://klipper.discourse.group/t/discussion-proposal-small-heater-control-improvements/19195)

I don't think PID is bad or needs to be replaced.
_I tried to do some fun stuff, like train DNN or write my own algo to control the power._
In the end, everything looks the same.

There are historically various topics, that try to replace it or tune it differently.
I think the only practical reason is to handle something important for that human being.

So, there is an ultimate solution, I think.
Pid is still under the control of a heater and is able to drive it.
The end-user has the ability to adjust it freely, like in my doc example - drive PWM in a feedforward way if the fan is suddenly enabled, like for bridges.

If it is needed, it is definitely possible to get the current chamber temperature, adjust to the fan curve, try to parse the extruder moving queue (IDK) and drive PWM higher before the temperature falls and PID will reactively compensate for that.
_I hope that will add enough flexibility to stop rituals like "how to do PID autotune properly"._

Another example, for some reason people do not insulate the heating beds, adding thermistors inside a thick aluminum plate and then want sort of PID control by two thermistors, and feel that combined sensor is not good enough.

Well, now it is possible to literally query the sensors and decide what to do, like return the `-N` value and decrease bed heater power on demand.

I hope it is made in a way to is safe and acceptable, leaves all existing safety measures in place, and mostly does not touch safety-critical code.

Thanks.

---
Btw, I will try to add templates to temperature FANs in a similar way, to fit most of the wishes, like curves, formulas & etc.

---
```
cd ~/klipper
git fetch origin pull/6837/head
git checkout FETCH_HEAD
sudo systemctl restart klipper
```

nefelim4ag committed:
heater_pc: define predictive control extras

Allows for the amendment of the real control loop
with feed-forward control

Feed-forward allows for the implementation of advanced control,
such as "dual loop", using data from several temperature sensors.
Increase/decrease power beforehand without PID reactive lag.

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

Sineos commented:
Nice!
It seems to follow the same logic as display templates. Maybe stick to the general wording and naming from there?

nefelim4ag commented:
>  Maybe stick to the general wording and naming from there?

I imported some changes from there, and I hope, now it is a little better.

Sineos commented:
From the eyes emoticon, I gather you found my comment confusing. What I tried to say:

The display template functions in Klipper for LEDs and fans ([example](https://www.klipper3d.org/Config_Reference.html#display_template)) seem to be using the same overarching logic, albeit with a bit of different nomenclature. 
Maybe I'm misunderstanding something?

nefelim4ag commented:
> The display template functions in Klipper for LEDs and fans ([example](https://www.klipper3d.org/Config_Reference.html#display_template)) seem to be using the same overarching logic, albeit with a bit of different nomenclature.

I got the idea of what you meant after your comment in the temperature_fan PR.
I'm looking for what can be done to make it similar to current integration with `[display_templates]'.
_The current integration feels a little unintuitive from my perspective, but it is better to reuse it, of course._

_Thanks for checking. I was totally unaware of that functionality._

nefelim4ag commented:
I don't know how I feel about adding a G-code command to reload the template or directly alter the PWM output. 
I also think it is wrong to add custom value get/set/log functions to all templates by incorporating them in `display_template`.
So, I dropped them for now.

So, for now, I only leave here a bare config binding between template and heater.
If it is needed, technically, temperature and read time can be propagated in the template context.
But I think they are mostly useless without state management inside macro.

Thanks.

KevinOConnor commented:
Interesting.  I'm not sure this implementation is safe though, as the PID `temperature_update()` code is called from a background thread (`serialhdl:_bg_thread()`) and I don't know what would happen if one tried to render templates from that thread.  Certainly if the template accessed any of the `{printer.module.xxx}` values it would result in `get_status()` calls from the background thread which would not be safe.

I suspect we could alter the target temperature from templates without thread issues (or set a temperature offset), but I'm not sure if that's what you are looking for.

Still interesting though.
-Kevin

nefelim4ag commented:
> code is called from a background thread (serialhdl:_bg_thread()) and I don't know what would happen if one tried to render templates from that thread

I will change it to run in a similar fashion to current fan/led templates in the main/reactor thread.
And pass shared value to offset PWM.
It seems simpler to do locks with one shared value.
The only tricky part would be to run it close before PWM calculation to make it semi-synchronized.
Maybe use the last PWM evaluation time and schedule the timer in <300ms after that.
Even though it will mostly retain the suggested functionality.

> I suspect we could alter the target temperature from templates without thread issues (or set a temperature offset), but I'm not sure if that's what you are looking for.

AFAIK, it is only useful if the target is to use variable extrusion temperature based on the flow. AFAIK, there is at least one project for that: https://github.com/sb53systems/G-Code-Flow-Temperature-Controller

It can be made on the Klipper side, but I suspect it is requires different design to trigger temperature changes ahead and do other stuff like accounting for thermal inertia/slow down printing.
So, I suspect it is easier to do as an external extension/gcode preprocessor.

Thanks.

nefelim4ag committed:
heaters: make control loop return value

This allows to chain the control later,
or reuse same control loops for the fan for example

Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>
